### PR TITLE
Make CDSDashboards.conda_envs dynamically update from function

### LINKED ISSUE #1293: [BUG] Available envs for dashboards is incorrect
### OS system and architecture in which you are running QHub

GCP

### Expected behavior

* all the envs should show up as options for deploying dashboards.

### Actual behavior

I have a group env on `https://quansight-beta.qhub.dev/` that I want to use as my dashboard env, but its not showing up in the list of available envs on the `Create a new dashboard` page. The environment _does_ show up under store `https://quansight-beta.qhub.dev/conda-store/environment/analyst/water_dashboards/` 

This is the same env that I have on `https://quansight.qhub.dev/` as a user env and on that deployment, it _does_ show up in the list of available envs when creating a new dashboard. 

### How to Reproduce the problem?

```channels:
- conda-forge
dependencies:
- python
- pip
- jupyterlab
- pandas
- holoviews
- geopandas
- geoviews
- param
- panel
- rich
- cdsdashboards-singleuser >=0.5.6
- jupyter_bokeh
- pip:
  - noaa-coops
- ipykernel
name: water_dashboards
prefix: null
```

### Command output

_No response_

### Versions and dependencies used.

https://quansight-beta.qhub.dev/ - QHub v0.4.1 according to the screen
https://quansight.qhub.dev/ - v0.0.1 according to the screen

### Compute environment

_No response_

### Integrations

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by costrouc**
Closes https://github.com/Quansight/qhub/issues/1310
Closes #1293

## Changes introduced in this PR:

- cdsdashboards conda environments are now populated by a function. In this case the environments will be updated based on the available environments from conda-store.

## Types of changes

What types of changes does your PR introduce?

_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

### Requires testing

- [X] Yes
- [ ] No

### In case you checked yes, did you write tests?

- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by costrouc** [comment]
I will update this issue later today. @viniciusdc will help with testing this PR once the tests pass.

**Comment by costrouc** [comment]
Will be working on this PR later tomorrow and have complete by Tuesday. Goal is to make use of the conda-store api to get available environments.

**Comment by trallard** [comment]
- @costrouc  will work on setting the service tokens for conda-store
- @HarshCasper  and @costrouc  will then work on fixing this 

---

Will be using the `conda-store` API

**Comment by costrouc** [comment]
@HarshCasper some details on how to accomplish this. 

In the [conda_store_config.py](https://github.com/Quansight/qhub/blob/main/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/conda-store/config/conda_store_config.py#L1) we will need to generate an api token via

```python
c.AuthenticationBackend.predefined_tokens = {
     f"{config["cdsdashboard-secret"]}": {
           "primary_namespace": "cdsdashboards",
           "role_bindings": {
                "*/*": ["viewer"],
           }
     }
}
```

This is passed inside of here https://github.com/Quansight/qhub/blob/main/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/conda-store/server.tf#L1-L23. The more I think about it these tokens and permissions should be passed in in https://github.com/Quansight/qhub/blob/main/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/conda-store/variables.tf#L70 as a variable with the secret token along with permissions needed.

[here](https://github.com/Quansight/qhub/blob/main/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/jupyterhub/files/jupyterhub/02-spawner.py#L52-L54) you will need to create a function which calls the conda-store kubernetes service api with this endpoint https://conda-store.readthedocs.io/en/latest/api.html#/api/api_list_environments_api_v1_environment__get. 

Getting started it might be easiest to hardcode the api token that is added to the conda-store config in order to test the api calls from cdsdashboards.

Later we can talk about how to link them in terraform code. We will likely do tricks similar to how this same pattern is handled in jupyterhub https://github.com/Quansight/qhub/blob/main/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/jupyterhub/variables.tf#L84-L87.

**Comment by viniciusdc** [comment]
When including the extra traitlet to condas_store_config.py I got the following error:
```
```

File: qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/jupyterhub/files/jupyterhub/02-spawner.py
Comment (viniciusdc) [inline qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/jupyterhub/files/jupyterhub/02-spawner.py:9]: Removing this line generated the following error:
```
 Loading extra config: 01-theme.py                                                                                                                                      │
│ Loading extra config: 02-spawner.py                                                                                                                                    │
│ [E 2022-09-05 18:46:15.302 JupyterHub app:2989]                                                                                                                        │
│     Traceback (most recent call last):                                                                                                                                 │
│       File "/opt/conda/lib/python3.7/site-packages/jupyterhub/app.py", line 2986, in launch_instance_async                                                             │
│         await self.initialize(argv)                                                                                                                                    │
│       File "/opt/conda/lib/python3.7/site-packages/jupyterhub/app.py", line 2477, in initialize                                                                        │
│         self.load_config_file(self.config_file)                                                                                                                        │
│       File "/opt/conda/lib/python3.7/site-packages/traitlets/config/application.py", line 110, in inner                                                                │
│         return method(app, *args, **kwargs)                                                                                                                            │
│       File "/opt/conda/lib/python3.7/site-packages/traitlets/config/application.py", line 889, in load_config_file                                                     │
│         raise_config_file_errors=self.raise_config_file_errors,                                                                                                        │
│       File "/opt/conda/lib/python3.7/site-packages/traitlets/config/application.py", line 844, in _load_config_files                                                   │
│         config = loader.load_config()                                                                                                                                  │
│       File "/opt/conda/lib/python3.7/site-packages/traitlets/config/loader.py", line 625, in load_config                                                               │
│         self._read_file_as_dict()                                                                                                                                      │
│       File "/opt/conda/lib/python3.7/site-packages/traitlets/config/loader.py", line 658, in _read_file_as_dict                                                        │
│         exec(compile(f.read(), conf_filename, "exec"), namespace, namespace)                                                                                           │
│       File "/usr/local/etc/jupyterhub/jupyterhub_config.py", line 446, in <module>                                                                                     │
│         exec(config_py)                                                                                                                                                │
│       File "<string>", line 5, in <module>                                                                                                                             │
│       File "/opt/conda/lib/python3.7/site-packages/kubespawner/__init__.py", line 13, in <module>                                                                      │
│         from kubespawner.spawner import KubeSpawner                                                                                                                    │
│       File "/opt/conda/lib/python3.7/site-packages/kubespawner/spawner.py", line 42, in <module>                                                                       │
│         from .objects import make_namespace                                                                                                                            │
│       File "/opt/conda/lib/python3.7/site-packages/kubespawner/objects.py", line 16, in <module>                                                                       │
│         from kubernetes.client.models import V1EndpointPort                                                                                                            │
│     ImportError: cannot import name 'V1EndpointPort' from 'kubernetes.client.models' (/opt/conda/lib/python3.7/site-packages/kubernetes/client/models/__init__.py) 
```

Which made the Hub deployment enter a crash loop cycle. Returning the block above fixed the problem.

**Comment by viniciusdc** [comment]
It needs to be tested with 'sha-563ef5c' for the docker images, waiting for 0.4.4 release.

**Comment by viniciusdc** [comment]
Blocked by https://github.com/ideonate/cdsdashboards/issues/107

**Comment by viniciusdc** [comment]
This is ready for review, when testing please be sure to use the latest available image from main

File: qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/conda-store/config/conda_store_config.py
Comment (costrouc) [inline qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/conda-store/config/conda_store_config.py:129]: We will need to modify the new property that will be exposed in this PR https://github.com/Quansight/conda-store/pull/388. `c.CondaStore.environment_directory = "/home/conda/{namespace}/envs/{namespace}-{name}"`. This will ensure that the conda environment name has the namespace embeded within it.

**Review by costrouc — COMMENTED** [review]
Will need to wait for a new release of Conda-Store which I will be releasing tonight.

File: qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/conda-store/config/conda_store_config.py
Comment (costrouc) [reply qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/conda-store/config/conda_store_config.py:129]: Bump the version to 0.4.12 as well

**Comment by viniciusdc** [comment]
I will do another drive-test on this, then we are set

**Comment by viniciusdc** [comment]
Tested, and all seems to be working!!!
Captura de tela de 2022-09-27 12 57 26
The new envs are dynamically updated in the dropdown menu once built on conda-store (`cdsdashboards` is required at least)

Captura de tela de 2022-09-27 12 59 16
 with the recent changes made to conda-store 0.4.12, we now have these weird names on the main page, but we can address this later.

**Comment by viniciusdc** [comment]
Thanks for updating the cypress tests @iameskild, once you approve, this is ready to be merged. Something that we will need to check once v0.4.5rc shows up in beta, is how conda-store old environments will behave with the name changes (we are modifying the exposed env names with this config  `c.CondaStore.environment_directory`)

**Review by iameskild — APPROVED** [review]
@viniciusdc I tested this myself and it works wonderfully :)

**Pull request merged**
